### ISSUE TITLE
S3 module doesn't use AWS Signature Version 4

### ISSUE THREAD

**Issue by Du-St** [ansible/ansible-modules-core#3140]
##### Issue Type:
- Bug Report
##### Ansible Version:

ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
##### Ansible Configuration:

NONE
##### Environment:
- CentOS 7 
- Ubuntu 14.04
##### Summary:

Using Ansible's S3 module I am not able to copy server-side encrypted files (Server Side Encryption with AWS KMS managed keys) from bucket to local directory although all needed settings are set and AWSCLI works well.

EC2 instance has IAM role with permissions to use the appropriate KMS key and access the bucket. I don't set variables access_key and access_secret_key explicitly. I am able to get non-encrypted files from the same bucket using Ansible's S3 module.
##### Steps To Reproduce:

CentOS7:
- yum install epel-release
- yum install python-pip
- yum install --enablerepo epel-testing ansible
- pip install --upgrade awscli
- aws configure set s3.signature_version s3v4
- ansible -vvv localhost -c local -m s3 -a 'region=eu-west-1 mode=get bucket=my-bucket object=/id_rsa dest=/root/.ssh/id_rsa'

But this works as expected:
- aws s3 cp 's3://my-bucket/id_rsa' /root/.ssh/id_rsa --region eu-west-1
##### Expected Results:

File id_rsa occurs in the /root/.ssh/ directory.
##### Actual Results:

Using /etc/ansible/ansible.cfg as config file
ESTABLISH LOCAL CONNECTION FOR USER: root
127.0.0.1 EXEC ( umask 22 && mkdir -p "$( echo $HOME/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705 )" && echo "$( echo $HOME/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705 )" )
127.0.0.1 PUT /tmp/tmppZN1Re TO /root/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705/s3
127.0.0.1 EXEC LANG=en_US.utf8 LC_ALL=en_US.utf8 LC_MESSAGES=en_US.utf8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705/s3; rm -rf "/root/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705/" > /dev/null 2>&1
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705/s3", line 2823, in <module>
    main()
  File "/root/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705/s3", line 496, in main
    download_s3file(module, s3, bucket, obj, dest, retries, version=version)
  File "/root/.ansible/tmp/ansible-tmp-1456127493.26-47301518663705/s3", line 323, in download_s3file
    key.get_contents_to_filename(dest)
  File "/usr/lib/python2.7/site-packages/boto/s3/key.py", line 1665, in get_contents_to_filename
    response_headers=response_headers)
  File "/usr/lib/python2.7/site-packages/boto/s3/key.py", line 1603, in get_contents_to_file
    response_headers=response_headers)
  File "/usr/lib/python2.7/site-packages/boto/s3/key.py", line 1435, in get_file
    query_args=None)
  File "/usr/lib/python2.7/site-packages/boto/s3/key.py", line 1467, in _get_file_internal
    override_num_retries=override_num_retries)
  File "/usr/lib/python2.7/site-packages/boto/s3/key.py", line 325, in open
    override_num_retries=override_num_retries)
  File "/usr/lib/python2.7/site-packages/boto/s3/key.py", line 273, in open_read
    self.resp.reason, body)
boto.exception.S3ResponseError: S3ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InvalidArgument</Code><Message>Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4.</Message><ArgumentName>Authorization</ArgumentName><ArgumentValue>null</ArgumentValue><RequestId>...</RequestId><HostId>...</HostId></Error>

localhost | FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_name": "s3"
    }, 
    "parsed": false
}

**Comment by jonbuffington** [comment]
We are experiencing the same error using Ansible 2.0.1.0 connecting to an Amazon Linux (EMR 4.5.0 release) that has both 2.39.0 and botocore 1.3.30 installed.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by egonbraun** [comment]
I am also having the same issue here when trying to download a file from an s3 in us-west-1 from an ec2 instance in Frankfurt. I am using Ansible Version 2.1.1.0 with Python 2.7.6 and boto 2.42.0.

```
- name: download_ssl_certificate
  s3: bucket="{{ s3.bucket }}"
      object="myfile"
      dest="{{ haproxy.ssl.certificate }}"
      region="{{ s3.region }}"
      aws_access_key="{{ s3.access_key }}"
      aws_secret_key="{{ s3.secret_key }}"
      mode=get
```

When running it I am having the error:

`fatal: [server]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "sudo: unable to resolve host ip-10-0-1-208\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_jog5IP/ansible_module_s3.py\", line 663, in <module>\r\n    main()\r\n  File \"/tmp/ansible_jog5IP/ansible_module_s3.py\", line 496, in main\r\n    download_s3file(module, s3, bucket, obj, dest, retries, version=version)\r\n  File \"/tmp/ansible_jog5IP/ansible_module_s3.py\", line 321, in download_s3file\r\n    key.get_contents_to_filename(dest)\r\n  File \"/usr/local/lib/python2.7/dist-packages/boto/s3/key.py\", line 1712, in get_contents_to_filename\r\n    response_headers=response_headers)\r\n  File \"/usr/local/lib/python2.7/dist-packages/boto/s3/key.py\", line 1650, in get_contents_to_file\r\n    response_headers=response_headers)\r\n  File \"/usr/local/lib/python2.7/dist-packages/boto/s3/key.py\", line 1482, in get_file\r\n    query_args=None)\r\n  File \"/usr/local/lib/python2.7/dist-packages/boto/s3/key.py\", line 1514, in _get_file_internal\r\n    override_num_retries=override_num_retries)\r\n  File \"/usr/local/lib/python2.7/dist-packages/boto/s3/key.py\", line 343, in open\r\n    override_num_retries=override_num_retries)\r\n  File \"/usr/local/lib/python2.7/dist-packages/boto/s3/key.py\", line 303, in open_read\r\n    self.resp.reason, body)\r\nboto.exception.S3ResponseError: S3ResponseError: 400 Bad Request\r\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n<Error><Code>InvalidArgument</Code><Message>Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4.</Message><ArgumentName>Authorization</ArgumentName><ArgumentValue>null</ArgumentValue><RequestId>1B519A08045695A2</RequestId><HostId>5Dk+LzfOqgz0m6RzxNumMvr+Zv/PN/qnFGmnQmwkqSywzO1dKMrJR9A8H62CdOTLldTQiQk9tYo=</HostId></Error>\r\n", "msg": "MODULE FAILURE", "parsed": false}`

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Du-St** [comment]
Hello Ansible maintainers!

Is there any progress?? When do you expect resolution of this issue?

Thanks.

D.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dtaylor-iidnet** [comment]
Hi Folks,

I have a workaround. Seeing as ansible uses boto under the covers, I looked for anyone talking about that there, and found this:
https://github.com/jschneier/django-storages/issues/28

BodhiSukha had a fix, involving setting the environment variable as follows:

```
os.environ['S3_USE_SIGV4'] = 'True'
AWS_S3_HOST = 's3.eu-central-1.amazonaws.com'
```

and leistungsabfall confirmed it. OK, they did it within a prog, so perhaps that is harder. I guess you could add it to the custom config for the user that uses ansible in boto, but that does not tally with boto config file(s) doco. I have tried setting the variable explicitly, before calling ansible, and it all works OK (where it did not before). So the steps now to reproduce a working workaround are:

CentOS7:

```
yum install epel-release
yum install python-pip
yum install --enablerepo epel-testing ansible
pip install --upgrade awscli
S3_USE_SIGV4=True
export S3_USE_SIGV4
ansible -vvv localhost -c local -m s3 -a 'region=eu-west-1 mode=get bucket=my-bucket object=/id_rsa dest=/root/.ssh/id_rsa'
```

Worked for me on ansible 2.3.0, boto 2.42.0, but then when I came to run it in a playbook, it went back to the old behaviour. I think this is because when running the playbook ansible was ssh-ing even to the local box where the task was running. I edited the task definition to include setting the env variable as follows:

```
- name: Get the Keystore file from the S3 bucket and store locally
  local_action: s3 bucket={{ my_bucket_var }} region={{ my_region_var}} object={{ my_cert_path_key_var }} dest={{ my_dest_dir_var }} mode=get
  environment:
    S3_USE_SIGV4: True
```

... obviously having set these variables elsewhere in the playbook. That finally worked.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ozbillwang** [comment]
+1 and we need a new option to nominate kms key which is not always using the default one.

**Comment by ryansb** [comment]
@SydOps please open a separate issue to track that need.

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/29872
